presto-hive/src/main/java/com/facebook/presto/hive/HiveTableStatisticsProvider.java

```diff
@@ -0,0 +1,103 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.hive;
+
+import com.facebook.presto.hive.metastore.ExtendedHiveMetastore;
+import com.facebook.presto.hive.metastore.Partition;
+import com.facebook.presto.hive.metastore.Table;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.ConnectorTableLayoutHandle;
+import com.facebook.presto.spi.connector.ConnectorTableStatisticsProvider;
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+import com.facebook.presto.spi.statistics.TableStatistics;
+import com.facebook.presto.spi.statistics.ValueEstimate;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.hive.HiveUtil.toPartitionValues;
+import static com.facebook.presto.hive.util.Types.checkType;
+import static com.facebook.presto.spi.statistics.TableStatistics.EMPTY_STATISTICS;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+
+public class HiveTableStatisticsProvider
+        implements ConnectorTableStatisticsProvider
+{
+    private final ExtendedHiveMetastore hiveMetastore;
+
+    @Inject
+    public HiveTableStatisticsProvider(ExtendedHiveMetastore hiveMetastore)
+    {
+        this.hiveMetastore = hiveMetastore;
+    }
+
+    @Override
+    public TableStatistics getTableStatistics(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorTableLayoutHandle layoutHandle)
+    {
+        HiveTableLayoutHandle layout = checkType(layoutHandle, HiveTableLayoutHandle.class, "layoutHandle");
+        List<HivePartition> hivePartitions = layout.getPartitions().orElse(ImmutableList.of());
+        List<Long> partitionRowNums = hivePartitions.stream()
+                .map(this::getPartitionStatistics)
+                .map(PartitionStatistics::getNumRows)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(toList());
+
+        long numRowsSum = partitionRowNums.stream().mapToLong(a -> a).sum();
+        long partitionsWithStatsCount = partitionRowNums.size();
+        long allPartitionsCount = hivePartitions.size();
+
+        TableStatistics tableStatistics = EMPTY_STATISTICS;
+        if (partitionsWithStatsCount > 0) {
+            tableStatistics = TableStatistics.builder()
+                    .setRowsCount(new ValueEstimate(1.0 * numRowsSum / partitionsWithStatsCount * allPartitionsCount))
```

Comment:
I think you should not treat partitions equally. I think it should be based on partition size.

Example
```
# partition | size | rows count
1 | 1000 | 10
2 | 500  | unknown
```
Would yield 15 rows instead of (in your case) 20.